File: node/src/auth/create.ts

import bcrypt from 'bcrypt';
import { PrismaClient } from '@prisma/client';

export interface ICreateUser {
  name: string,
  email: string,
  password: string,
}

export async function createUser(db: PrismaClient, data: ICreateUser): Promise<{ sessionID: string, userID: string } | null> {
  let user = await db.user.findFirst({
    where: {
      email: data.email,
    }
  });

  if (user) {
    return null;
  }

  const hashedPass = bcrypt.hashSync(data.password, 10);
  user = await db.user.create({
    data: {
      name: data.name,
      email: data.email,
      password_hash: hashedPass,
    },
  });

  const session = await db.session.create({
    data: { user_id: user.id },
  });

  return {
    sessionID: session.id,
    userID: user.id,
  };
}

File: rust/src/auth/create.rs

// ===== Imports =====
use anyhow::{Result, anyhow};
use sea_orm::{DatabaseConnection, EntityTrait, QueryFilter, Set, ColumnTrait};
use serde::{Serialize, Deserialize};
// ===================

#[derive(Serialize, Deserialize)]
pub struct CreateUserPayload {
  name: String,
  email: String,
  password: String,
}

pub async fn create_user(db: &DatabaseConnection, user: CreateUserPayload) -> Result<(String, String)> {
  let users = entity::user::Entity::find()
    .filter(entity::user::Column::Email.eq(user.email.clone()))
    .all(db).await?;
  if !users.is_empty() {
    return Err(anyhow!("User with given email already exists"));
  }

  let user_id = uuid::Uuid::new_v4().to_string();
  let password_hash = bcrypt::hash(user.password, 10)?;
  let user_doc = entity::user::ActiveModel {
    id: Set(user_id.clone()),
    name: Set(user.name),
    email: Set(user.email.clone()),
    password_hash: Set(password_hash),
  };
  entity::user::Entity::insert(user_doc).exec(db).await?;

  let session_id = uuid::Uuid::new_v4().to_string();
  let session_doc = entity::session::ActiveModel {
    id: Set(session_id.clone()),
    user_id: Set(user_id.clone()),
  };
  entity::session::Entity::insert(session_doc).exec(db).await?;

  Ok((session_id, user_id))
}